spmd/tensor/parallel/api.py

```diff
@@ -0,0 +1,100 @@
+    row_wise_sharding: Sequence[Placement] = [Shard(1)]
+    replicate: Sequence[Placement] = [Replicate()]
+
+    def _shard_self_attn_params(name: str, module: nn.Module) -> None:
+        if isinstance(module, nn.Linear):
+            if name == "qkv":
+                sharded_weight = nn.Parameter(
+                    distribute_tensor(
+                        module.weight, device_mesh, col_wise_sharding
+                    )
+                )
+                module.register_parameter("weight", sharded_weight)
+                if module.bias is not None:
+                    sharded_bias = nn.Parameter(
+                        distribute_tensor(
+                            module.bias, device_mesh, col_wise_sharding
+                        )
+                    )
+                    module.register_parameter("bias", sharded_bias)
+            elif name == "proj":
+                sharded_weight = nn.Parameter(
+                    distribute_tensor(
+                        module.weight, device_mesh, row_wise_sharding
+                    )
+                )
+                module.register_parameter("weight", sharded_weight)
+                if module.bias is not None:
+                    replicated_bias = nn.Parameter(
+                        distribute_tensor(module.bias, device_mesh, replicate)
+                    )
+                    module.register_parameter("bias", replicated_bias)
+
+    if isinstance(module, TensorParallelMultiheadAttention):  # shard TPMA
+        for n, m in module.named_children():
+            _shard_self_attn_params(n, m)
+    else:
+        for n, m in module.named_children():  # replace with TPMA
+            if isinstance(m, nn.MultiheadAttention):
+                tp_multi_head_attention = TensorParallelMultiheadAttention(
+                    m.embed_dim,
+                    m.num_heads,
+                    device=torch.device(
+                        device_mesh.device_type
+                    ),  # TODO: convert DeviceMesh.device_type from str to torch.device???
+                    tp_size=device_mesh.size(0),  # group size on dim 0
+                    add_bias_kv=m.bias_k is not None,
+                )
+                tp_multi_head_attention.copy(m)
+                module.register_module(n, tp_multi_head_attention)
+
+
+def tp_shard_self_attn(
+    device_mesh: DeviceMesh,
+) -> Callable[[str, nn.Module], None]:
+    return functools.partial(_shard_self_attn, device_mesh=device_mesh)
+
+
+def replicate_input(
+    device_mesh: DeviceMesh,
+) -> functools.partial[Tuple[DTensor, ...]]:
+    return functools.partial(_replicate_input, device_mesh=device_mesh)
+
+
+def aggregate_output(output: DTensor) -> torch.Tensor:
+    if isinstance(output, DTensor):
+        replica_placement = [Replicate()]
+        return (
+            output.redistribute(output.device_mesh, replica_placement)
+            .to_local()
+            .contiguous()
```

Comment:
do we really need the output be contiguous?

Reply:
Good question. I created a new task to later investigate this but I'll keep it as `contiguous` in this PR.